controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -483,6 +603,87 @@ func metricLabels(endpoints *corev1.Endpoints) map[string]string {
 	return labels
 }
 
+func (pp *portPublisher) endpointSliceToAddresses(es *dv1beta1.EndpointSlice) AddressSet {
+	addresses := make(map[ID]Address)
+
+	resolvedPort := pp.resolveESTargetPort(es.Ports)
+	for _, endpoint := range es.Endpoints {
+		if pp.hostname != "" && pp.hostname != *endpoint.Hostname {
+			continue
+		}
+
+		// Without this we get `Add:` twice, once when !ready
+		// and once when ready
+		if !*endpoint.Conditions.Ready {
+			continue
+		}
+
+		// Every endpoint has addresses that correspond to the EndpointSlice addressType field
+		// Consumers must handle different types of addresses
+		// An "addresses" set must contain at least 1 but no more than 100 addresses
+		if endpoint.TargetRef == nil {
+			serviceName := es.Labels[dv1beta1.LabelServiceName]
+			for _, addr := range endpoint.Addresses {
+				if isIPv6Address(addr) {
+					continue
+				}
+
+				id := ServiceID{
+					Name: strings.Join([]string{
+						serviceName,
+						addr,
+						fmt.Sprint(resolvedPort),
+					}, "-"),
+					Namespace: es.Namespace,
+				}
+
+				var authorityOverride string
+				if fqName, ok := es.Annotations[consts.RemoteServiceFqName]; ok {
+					authorityOverride = fmt.Sprintf("%s:%d", fqName, pp.srcPort)
+				}
+				addresses[id] = Address{
+					IP:                addr,
+					Port:              resolvedPort,
+					Identity:          es.Annotations[consts.RemoteGatewayIdentity],
+					AuthorityOverride: authorityOverride,
+				}
+			}
+			continue
+		}
+
+		if endpoint.TargetRef.Kind == "Pod" {
+			for _, addr := range endpoint.Addresses {
+				if isIPv6Address(addr) {
```

Comment:
I think instead of doing so you can simply check the `AddressType` on the EndpointsSlice object. A slice will only have addresses of one type in its endpoints unless the documentation is incorrect.